### DISCUSSION TITLE
Unable to get custom filters work in server driven datagrid

### DISCUSSION THREAD

**Discussion by laughterborn** [vmware/clarity#5621] [Q&A]
Hi,
I'm planning to have custom filters for columns in server driven datagrid. However,I've found too few references/examples to achieve the same.
I want to know, if at all the customer filters are supported with server driven datagrid and is there any pointer for the same.

Thanks in advance!

**Comment by hippee-lee** [comment]
Great question @laughterborn - we have some documentation and code examples [here](https://clarity.design/angular-components/datagrid/#custom-filtering). If you want to see a working example you can refer to the internal dev app [here](https://github.com/vmware/clarity/tree/next/packages/angular/projects/dev) and the custom filter demos [here](https://github.com/vmware/clarity/tree/next/packages/angular/projects/dev/src/app/datagrid/utils).

**Reply by laughterborn** [reply]
Thanks @hippee-lee, I'd seen these examples and it don't contain usage of custom filter in pagination.
Also,I'm interested in understanding the usage of (clrDgRefresh)in component. Please point me to any other detailed documentation which explains this(If any).

**Reply by hippee-lee** [reply]
`clrDgRefresh` is for a[ server driven datagrid](https://clarity.design/angular-components/datagrid/#server-driven-datagrid). 

In that case you would pass the filter back to your server so it could filter the paginated dataset. Most folks use the examples on that page to get started but the internal dev app demo [here](https://github.com/vmware/clarity/tree/next/packages/angular/projects/dev/src/app/datagrid/server-driven) might also help.

If thats not enough, please post a [stackblitz](http://stackblitz.com/@clr-team) with a simple example of what you are trying so we can look at your code and offer advice.

**Comment by laughterborn** [comment]
@hippee-lee  ,Please find the stackblitz example here https://stackblitz.com/edit/clarity-datagrid-server-driven?embed=1&file=src/app/app.module.ts

In this I want to have a custom filter for sorting the column "year" with (>,< and =) operators.
Another observation is, state is not listening to the filter changes when applied.

**Comment by kumar-tadepalli** [comment]
Any update on this?

**Comment by laughterborn** [comment]
Hi @hippee-lee , any update on the above question?

**Reply by hippee-lee** [reply]
@laughterborn and @shravansofts - I extracted the color filter demo from the website and our Angular dev app into a standalone app for an easier to understand example of server-driven, pagination and custom filtering. 

[This repo](https://github.com/hippee-lee/clarity-angular-demos) uses the latest (as of the HEAD commit) angular and Clarity libs. Stackblitz has been problematic so for this case I just created a seperate repo that can be cloned, installed and run on a local machine.

Take a look at [app.component.ts](https://github.com/hippee-lee/clarity-angular-demos/blob/next/src/app/app.component.ts) for the application template to see how it incorporates the custom filter, pagination and refresh function for server driven datagrids.
Take a look at [color-filter.ts](https://github.com/hippee-lee/clarity-angular-demos/blob/next/src/app/utils/color-filter.ts) to see how it implements the custom filtering logic and UI controls for the filter. 

FYI [inventory.ts](https://github.com/hippee-lee/clarity-angular-demos/blob/next/src/app/utils/inventory.ts) is a mock backend so it will likely be different from the apps backend. The thing to know is that it it is filtering `on the server` by passing the color filter in its current state applied to the dataset. I'm calling this out because the code inside `refresh` function is specific to the mock inventory backend so the app you are working with is going to be different.

The greaterThan / LessThan filter described above should be implemented in a similar manner. You will have to implement the UI template that works with the Filter logic and then the refresh logic specific to the apps server side filtering API will take care of the rest. I hope this is enough to help you move forward.

**Reply by laughterborn** [reply]
Thanks @hippee-lee for this example, I'll clone this project and build it on my local. 

Also, I would want to reset the filter and sort on context change , So I'm using "**@ViewChildren(ClrDatagridColumn) columns: QueryList<ClrDatagridColumn>; this.columns.forEach(column => column.filterValue = "");** ". This leaves the filter selected.Please let me know, if there's a better approach.

**Reply by hippee-lee** [reply]
Qq - is the context change coming from inside the datagrid or is it external?

If it's inside the datagrid (filter/sorting change events) the datagrid emits the output `clrDgRefresh` whenever data needs to be refreshed based on user action. 

But if it's an external actions you can just call the method refresh calls to update the data. 

Does that make help?

**Reply by hippee-lee** [reply]
Sorry, I just re-read my answer and I misunderstood the question. 

If you want to reset the sorting you can use something like:
```html
  <clr-dg-column [clrDgField]="'pokemon.name'" [clrDgSortBy]="pokemonComparator" [(clrDgSortOrder)]="sortOrder"
    >Pokemon</clr-dg-column
```

where `sortOrder` is one value of the ClrDatagridSortOrder enum:

```typescript
export enum ClrDatagridSortOrder {
  UNSORTED = 0,
  ASC = 1,
  DESC = -1,
}
```
Filtering is similar but you need to use the `clrFilterValue`. E.g -
```html
  <clr-dg-column>
    Pokemon
    <clr-dg-string-filter
      *ngIf="displayFilter"
      [clrDgStringFilter]="pokemonFilter"
      [(clrFilterValue)]="myFilterValue"
    ></clr-dg-string-filter>
  </clr-dg-column>
```

where the value of the variable `myFilterValue` is set on the application side. This is two-way binding so it also reflects both values set by the application and if the user types in the filter input box. 

I know the datagrid is complicated but these use cases are covered in the demo code sections of the [datagrid docs](https://clarity.design/angular-components/datagrid/#code-examples).

**Reply by laughterborn** [reply]
Hi @hippee-lee , Thanks for the reply.

I built your example on local and it was working with pagination and filtering until I added _(clrDgRefresh)="refresh($event)"_ to the <clr-datagrid> referring the following documentation. https://clarity.design/angular-components/datagrid/#server-driven-datagrid

Also, I have replaced smart iterator (*clrDgItems) with ngFor . Adding these 2 changes made previous and next button unavailable. 

The following has the changes for your code. Could you please specify if I'm missing something in the implementation.
https://github.com/laughterborn/clarity-angular-demos-next

**Reply by hippee-lee** [reply]
Hi @laughterborn - it looks like there may be an issue with me copying/pasting the inventory mock server code. I'm sorry but I don't have time to dig into it today and it does still work in the dev app. IMO - its not worth investigating because inventory service was created as a quick and simple mock and not intended to be prescriptive in developing a server driven datagrid. 

I would suggest taking the ideas here and applying them to a real server driven API. The backend and refresh code will depend on the apps backend API.

**Reply by laughterborn** [reply]
Hi @hippee-lee,
 I understand that backend will handle the data processing and we have that in place, but in order to build the query params for calling the backend API from UI.If I've understand properly, datagrid should provide the applied filters and sorting state (ClrDatagridStateInterface) on change right? **which is not happening using custom filter.**

As per the clarity documentation, the server driven datagrid will be acheived by using (clrDgRefresh)="refresh($event)" on datagrid. In your example, I noticed it is not used.

We are confused and blocked with implementation.It would be great,If you could please confirm whether custom filters will be supported using server driven paginated datagrid or not?

**Reply by hippee-lee** [reply]
Custom filters have always been provided when they are one of the active filters in the ClrDatagridStateInterface. They were part of the first version of datagrid release in November 2016. 
The state object has always worked like this: 

- If any datagrid filters are  are active, there is an array of the active filters in the state object. 
- **It is the responsibility of the application to handle active filters in a way that makes sense for the backend before the application makes the request.** 

Please look [here](https://github.com/hippee-lee/clarity-angular-demos/commit/594623de073607a9a5be9d7f5158d692c56ed40f#diff-884f7f49640e5923f6bcac4c51d90340330a178f662defbe61e5f5aac1c512deR30) where I updated the demo to use `*ngFor` and modified the refresh function to provide selected colors info to the Mock inventory API / backend. 

Here is a capture of the working implementation.
![5621](https://user-images.githubusercontent.com/433692/115056297-7da2be80-9e97-11eb-9342-8fb19a621b69.gif)

**I have mentioned this several times but its worth repeating.** This implementation is specific to the dev application from several years ago. I've extracted and modified it rather than coming up with a new custom filter server driven datagrid demo. The Inventory service is a mock to approximate a fake backend and the code I wrote to utilize the custom color filter is specific to both the color filter and the Inventory service. For your application the custom filter + the refresh function code needs to conform the the API that is being hooked up to the datagrid.

**Reply by kumar-tadepalli** [reply]
@hippee-lee What is the use of accepts() method, in case of server driven pagination. I'm not seeing this method being called while applying to filters in case of server driven pagination.

I could not find any explanation on significance of all properties/methods in the below interface:
interface ClrDatagridFilterInterface<T, S=any> {
    isActive(): boolean;
    accepts(item: T): boolean;
    changes: Observable<any>;
    readonly state?: S;
    equals?(other: ClrDatagridFilterInterface<T, any>): boolean;
}

**Reply by hippee-lee** [reply]
It will tell the backend what to filter on. E.g  in the state object emitted it will pass which model property and what the current value for filtering is so that the backend can do its work and return the correct data set.
